llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp

```diff
@@ -0,0 +1,2348 @@
+                                          LVType *IndexType) {
+  assert(Array && "Invalid logical element");
+  assert(GSR && "Invalid metadata node.");
+  LLVM_DEBUG({
+    dbgs() << "\n[constructGenericSubrange] DIGenericSubrange\n";
+    GSR->dump();
+  });
+
+  LLVM_DEBUG({ dbgs() << "\nNot implemented\n"; });
+}
+
+// DIImportedEntity
+void LVIRReader::constructImportedEntity(LVElement *Element,
+                                         const DIImportedEntity *IE) {
+  assert(Element && "Invalid logical element");
+  assert(IE && "Invalid metadata node.");
+  LLVM_DEBUG({
+    dbgs() << "\n[constructImportedEntity] DIImportedEntity\n";
+    IE->dump();
+  });
+
+  if (LVElement *Import = constructElement(IE)) {
+    Import->setIsFinalized();
+    addSourceLine(Import, IE);
+    LVScope *Parent = getParentScope(IE);
+    Parent->addElement(Import);
+
+    const DINode *Entity = IE->getEntity();
+    LVElement *Target = getElementForSeenMD(Entity);
+    if (!Target) {
+      if (const auto *Ty = dyn_cast<DIType>(Entity))
+        Target = getOrCreateType(Ty);
+      else if (const auto *SP = dyn_cast<DISubprogram>(Entity))
+        Target = getOrCreateSubprogram(SP);
+      else if (const auto *NS = dyn_cast<DINamespace>(Entity))
+        Target = getOrCreateNamespace(NS);
+      else if (const auto *M = dyn_cast<DIModule>(Entity))
+        Target = getOrCreateScope(M);
+    }
+    Import->setType(Target);
+  }
+}
+
+LVScope *LVIRReader::getOrCreateAbstractScope(LVScope *OriginScope,
+                                              const DILocation *DL) {
+  assert(OriginScope && "Invalid logical element");
+  assert(DL && "Invalid metadata node.");
+  LLVM_DEBUG({
+    dbgs() << "\n[getOrCreateAbstractScope] DILocation\n";
+    DL->dump();
+  });
+
+  const DILocation *InlinedAt = DL->getInlinedAt();
+  DILocalScope *Context = DL->getScope();
+  LLVM_DEBUG({
+    dbgs() << "\nParent Scope:\n";
+    OriginScope->getParentScope()->dump();
+    dbgs() << "\nOriginScope:\n";
+    OriginScope->dump();
+    dbgs() << "\nInlinedAt:\n";
+    InlinedAt->dump();
+    dbgs() << "\nContext:\n";
+    Context->dump();
+  });
+
+  dwarf::Tag Tag = OriginScope->getTag();
+  if (OriginScope->getIsFunction() || OriginScope->getIsInlinedFunction()) {
+    Tag = dwarf::DW_TAG_inlined_subroutine;
+    OriginScope->setInlineCode(dwarf::DW_INL_inlined);
+  }
+  LVScope *AbstractScope = static_cast<LVScope *>(createElement(Tag));
+  if (AbstractScope) {
+    addInlinedScope(OriginScope, AbstractScope);
+    AbstractScope->setTag(Tag);
+    AbstractScope->setIsFinalized();
+    AbstractScope->setName(OriginScope->getName());
+    AbstractScope->setType(OriginScope->getType());
+
+    AbstractScope->setCallLineNumber(InlinedAt->getLine());
+    AbstractScope->setCallFilenameIndex(
+        getOrCreateSourceID(InlinedAt->getFile()));
+
+    AbstractScope->setReference(OriginScope);
+    AbstractScope->setHasReferenceAbstract();
```

Comment:
The object name is `AbstractScope`, but you're setting the inlining-site information on it and referring to the "origin" scope from it. This really sounds like the "Abstract" scope is actually the _inlined_ scope. An inlined subroutine in DWARF is what contains that information:

```
0x0000008a:     DW_TAG_inlined_subroutine
                  DW_AT_abstract_origin (0x00000036 "Max")
                  DW_AT_low_pc  (0x000000000000000b)
                  DW_AT_high_pc (0x0000000000000014)
                  DW_AT_call_file       ("/home/jmorse/test2.c")
                  DW_AT_call_line       (16)
                  DW_AT_call_column     (0x0c)
```
i.e. it has a call line, file index, and reference to another scope.